python/cudf/cudf/tests/test_doctests.py

```diff
@@ -0,0 +1,66 @@
+import doctest
+import inspect
+import os
+
+import numpy as np
+import pytest
+
+import cudf
+
+
+def _name_in_all(parent, name, member):
+    return name in getattr(parent, "__all__", [])
+
+
+def _is_public_name(parent, name, member):
+    return not name.startswith("_")
+
+
+def _find_docstrings_in_obj(finder, obj, criteria=None):
+    for docstring in finder.find(obj):
+        if docstring.examples:
+            yield docstring
+    for name, member in inspect.getmembers(obj):
+        # Filter out non-matching objects with criteria
+        if criteria is not None and not criteria(obj, name, member):
+            continue
+        # Recurse over the public API of modules (objects defined in __all__)
+        if inspect.ismodule(member):
+            yield from _find_docstrings_in_obj(
+                finder, member, criteria=_name_in_all
+            )
+        # Recurse over the public API of classes (attributes not prefixed with
+        # an underscore)
+        if inspect.isclass(member):
+            yield from _find_docstrings_in_obj(
+                finder, member, criteria=_is_public_name
+            )
+
+
+def _fetch_doctests():
+    finder = doctest.DocTestFinder()
+    yield from _find_docstrings_in_obj(finder, cudf, criteria=_name_in_all)
+
+
+class TestDoctests:
+    @pytest.fixture(autouse=True)
+    def chdir_to_tmp_path(cls, tmp_path):
+        original_directory = os.getcwd()
+        try:
+            os.chdir(tmp_path)
+            yield
+        finally:
+            os.chdir(original_directory)
```

Comment:
Did we figure out the below is sufficient? Codes after yield is run at tear down stage.
```suggestion
        os.chdir(tmp_path)
        yield
        os.chdir(original_directory)
```

Reply:
This try/finally guards against something throwing in the test (during the “yield”). I am not 100% sure about this but I suspect it is a good idea, from past work with similarly yielding functions.

Reply:
I don't feel too strongly about this so feel free to ignore. I did try the following piece that shows if the code inside the test goes very wrong and kills the process, the above won't be able to guard it.

```python
@pytest.fixture
def some_obj():
    try:
        print("setup obj")
        obj = ["object"]
        yield obj
    finally:
        print("tear down code")

def test_setup_teardown(some_obj):
    os.kill(os.getpid(), signal.SIGUSR1)
```

Reply:
I was hoping to guard against lesser evils than what’s in your example. 😂 But I suppose a “throw” would be caught by pytest… so I’m not sure. Perhaps it is fine to remove the try/finally here.

Reply:
Solid example @vyasr. Changed in d6553db05889e9e10ef6e0ded2db88c038903ae5.